modules/dag-l0/src/main/scala/io/constellationnetwork/dag/l0/http/routes/DelegatedStakesRoutes.scala

```diff
@@ -140,6 +147,144 @@ final case class DelegatedStakesRoutes[F[_]: Async: Hasher](
+
+    } yield
+      RewardsInfo(
+        epochsPerYear = emConfig.epochsPerYear,
+        currentDagPrice = currentPrice,
+        nextDagPrice = nextPrice,
+        totalDelegatedAmount = totalDelegateStake,
+        latestAverageRewardPerDag = avgRewardAmount,
+        totalDagAmount = currentTotalSupply,
+        totalRewardPerEpoch = latestDelegateRewardsNoCommission,
+        totalRewardsPerYearEstimate = totalRewardsPerYear
+      )
+
+  private def processDelegations(info: GlobalSnapshotInfo): F[(Amount, Amount, Amount, Amount)] = {
+    val activeDelegatedStakes = info.activeDelegatedStakes
+      .getOrElse(SortedMap.empty[Address, List[DelegatedStakeRecord]])
+
+    val pendingWithdrawals = info.delegatedStakesWithdrawals
+      .getOrElse(SortedMap.empty[Address, List[PendingDelegatedStakeWithdrawal]])
+
+    val totalSpendableSupply = info.balances.values.map(_.value.value).sum
+    val totalPendingSupply = pendingWithdrawals.values.flatten.map(_.rewards.value.value).sum
+
+    val (totalStakeLocked, totalActiveRewards) = activeDelegatedStakes.values.flatten.foldLeft((0L, 0L)) {
+      case ((stakeAcc, rewardsAcc), record) =>
+        val stakeAmount = record.event.value.amount.value.value
+        val rewardsAmount = record.rewards.value
+        (stakeAcc + stakeAmount, rewardsAcc + rewardsAmount)
+    }
+
+    (
+      toAmount(totalStakeLocked),
+      toAmount(totalActiveRewards),
+      toAmount(totalStakeLocked + totalActiveRewards), // totalDelegateStake
+      toAmount(totalSpendableSupply + totalPendingSupply + totalActiveRewards) // currentTotalSupply
+    ).mapN(Tuple4.apply)
+  }
+
+  private def getLatestDelegateRewardTotal(snapshot: GlobalIncrementalSnapshot, info: GlobalSnapshotInfo): F[Amount] = {
+    val delegateRewards = snapshot.delegateRewards
+      .getOrElse(SortedMap.empty[PeerId, Map[Address, Amount]])
+
+    val nodeParams = info.updateNodeParameters
+      .getOrElse(SortedMap.empty[ID.Id, (Signed[UpdateNodeParameters], SnapshotOrdinal)])
+
+    val calcFullReward: (Long, (PeerId, Map[Address, Amount])) => Long = {
+      case (acc, (peerId, rewards)) =>
+        val nodeCommission = BigDecimal(
+          nodeParams.get(peerId.toId).map(_._1.delegatedStakeRewardParameters.reward).getOrElse(0.0)
+        )
+
+        val delegatePortion = 1 - nodeCommission
+
+        acc + (BigDecimal(rewards.values.map(_.value.value).sum) / delegatePortion)
+          .setScale(0, RoundingMode.HALF_UP)
+          .longValue
+    }
+
+    delegateRewards
+      .foldLeft(0L)(calcFullReward)
+      .pure[F]
+      .flatMap(toAmount)
+  }
+
+  private def calculateAverageReward(latestRewards: Amount, totalStakedAmount: Amount): F[Amount] =
+    if (totalStakedAmount.value.value === 0) Amount.empty.pure[F]
+    else
+      toAmount(
+        (BigDecimal(latestRewards.value.value) / BigDecimal(totalStakedAmount.value.value))
```

Comment:
Similar thing, it'd be great if we can hide the decimal interactions behind an API we control so it's more strictly verifiable.